Implement utility traits for tendermint data types

yihuang opened the pull request:
I removed the ``Default`` instances that don't make sense.

tarcieri commented on tendermint/src/account.rs:
```diff
@@ -15,7 +15,7 @@ use subtle_encoding::hex;
 /// Account IDs
-#[derive(Copy, Clone, Eq, Hash, PartialEq, PartialOrd, Ord)]
+#[derive(Copy, Clone, Eq, Hash, PartialEq, PartialOrd, Ord, Default)]
 pub struct Id([u8; LENGTH]);
```
This was the main one I said didn’t make sense 😜 (but also anything with a hash)

tarcieri commented on tendermint/src/public_key.rs:
```diff
@@ -91,6 +91,13 @@ impl PublicKey {
+impl Default for PublicKey {
+    fn default() -> PublicKey {
+        static DUMMY_BYTES: [u8; 32] = [0; 32];
+        PublicKey::Ed25519(ed25519::PublicKey::new(DUMMY_BYTES))
```
This also doesn’t make any sense, and is a particularly pathological case as it’s a special point on the elliptic curve.

tarcieri commented on tendermint/src/time.rs:
```diff
@@ -36,13 +38,38 @@ impl Time {
 
+    /// Returns an RFC 3339 string, such as 1996-12-19T16:39:57-08:00.
+    pub fn to_rfc3339(&self) -> String {
+        self.0.to_rfc3339()
```
If you want to support this, it should be Z-normalized, which I think it is but that shiuld be reflected in the comment.

yihuang commented:
* Removed ``Default`` instance for ``Account::Id``, ``PublicKey``, ``validator::Info``.
* Fixed ``Time`` format to use ``.to_rfc3339_opts(SecondsFormat::Micros, true)``

tarcieri commented on tendermint/src/abci/transaction/hash.rs:
```diff
@@ -28,6 +29,15 @@ impl Hash {
+impl Default for Hash {
+    fn default() -> Hash {
+        // hash of empty data
+        let mut bytes = [0u8; LENGTH];
```
A `Hash` of all zero bytes is generally not semantically meaningful and IMO should not be the `Default`.

(I have seen security vulnerabilities arising from zero-value hashes in Go)

yihuang replied:
Makes sense, just removed.

tarcieri commented on tendermint/src/time.rs:
```diff
@@ -36,13 +38,38 @@ impl Time {
 
+impl Default for Time {
+    fn default() -> Time {
+        Time::unix_epoch()
```
I'm not sure a "default time" makes sense

yihuang replied:
Makes sense, just removed.

yihuang commented:
I've changed the format of `Time` to `SecondsFormat::Nanos`, according to [tendermint implementation](https://github.com/tendermint/tendermint/blob/master/types/canonical.go#L13).

tarcieri commented on tendermint/src/node/info.rs:
```diff
@@ -58,10 +58,15 @@ pub struct ProtocolVersionInfo {
 
 /// Listen address information
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, Default)]
```
It seems like this could have a more useful default (like, say, `0.0.0.0`, or whatever Tendermint does)

yihuang replied:
Agree, I think the default is "tcp://127.0.0.1:26656";

yihuang replied:
But, again, a concret network address might not suit for `Default` trait. Maybe I remove this instance.

tarcieri commented on tendermint/src/hash.rs:
```diff
@@ -28,6 +28,12 @@ pub enum Hash {
 
+impl Default for Hash {
+    fn default() -> Hash {
+        Hash::Null
```
Does this actually make sense? Honestly I think this whole concept should be replaced with `Option<Hash>` instead...

yihuang replied:
Agree, but that probably should belong to another PR?

tarcieri commented on tendermint/src/node/info.rs:
```diff
@@ -33,7 +33,7 @@ pub struct Info {
 
 /// Protocol version information
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, Default)]
```
Are zeroes for all of these version values semantically meaningful?

yihuang replied:
Removed.

tarcieri commented on tendermint/src/version.rs:
```diff
@@ -2,7 +2,7 @@ use serde::{Deserialize, Serialize};
 
 /// Tendermint version
-#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq)]
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Default)]
```
Should the default version really be empty string?

yihuang replied:
Removed.

yihuang commented on tendermint/src/block/header.rs:
```diff
@@ -74,7 +74,7 @@ pub struct Header {
 ///
 /// <https://github.com/tendermint/tendermint/blob/master/docs/spec/blockchain/blockchain.md#version>
-#[derive(Serialize, Deserialize, Clone, Debug)]
+#[derive(Serialize, Deserialize, Clone, Debug, Default)]
```
I think `Default` supposed to some universal default value for the data type, like unit in a monoid, should not change any time in the future (it would be a breaking change).

yihuang replied:
Removed.

yihuang commented on tendermint/src/block/id.rs:
```diff
@@ -16,7 +16,7 @@ pub const PREFIX_LENGTH: usize = 10;
 ///
 /// <https://github.com/tendermint/tendermint/blob/master/docs/spec/blockchain/blockchain.md#blockid>
-#[derive(Serialize, Deserialize, Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
+#[derive(Serialize, Deserialize, Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord, Default)]
```
`block::Id` support empty value, that's a good candidate for a `Default` value.

yihuang commented:
> Perhaps I'm a bit jaded by (in some cases security) vulnerabilities relating to Go zero values, which it seems is how `Default` is being leveraged in the cases I pointed out in this review.
> 
> If you're looking for a way to easily construct test fixtures, I think `Default` is the wrong tool for the job. Alternatively, in the test suite itself, you can define a trait _specifically for test fixtures_ in the _test suite itself_, and then impl it however you want!
> 
> ```rust
> trait TestFixture {
>     fn test_fixture() -> Self;
> }
> 
> impl TestFixture for Hash {
>     fn test_fixture() -> Hash {
>         Hash::Null
>     }
> }
> ```
> 
> In general, I think it's a bad idea for anything that's a testing-related convenience to leak outside of the test context, especially when it could potentially lead to security issues in production.
> 
> Is that what most of these are for?

Yeah, actually we've changed the test code to use more properly constructed mocking data.
But what's left in this patch now IMO are quite reasonable ;D

tarcieri commented:
@yihuang looks like this needs a GPG signature

yihuang commented:
> @yihuang looks like this needs a GPG signature

Done.
